import { SQSEvent } from 'aws-lambda'
import { DynamoDB, SQS } from 'aws-sdk'
import { RestoreMessage } from './types'
import { StreamSpecification } from 'aws-sdk/clients/dynamodb'

export const handler = async (event: SQSEvent): Promise<void> => {
  try {
    const sqs = await new SQS()
    const dynamoDB = await new DynamoDB()
    for (let i = 0; i < event.Records.length; i++) {
      let message = JSON.parse(event.Records[i].body) as RestoreMessage
      if (message.status === 'COMPLETED') continue
      if (!tableExists(message.from_table, dynamoDB)) {
        await dynamoDB
          .restoreTableToPointInTime({
            SourceTableName: message.to_table,
            TargetTableName: message.from_table,
            UseLatestRestorableTime: true,
          })
        message = { ...message, from_table: message.to_table, to_table: message.from_table, status: 'RESTORING_TO_ORIGINAL_TABLE' }
      } else if (tableExists(message.to_table, dynamoDB) && message.status !== 'DELETING_ORIGINAL_TABLE') {
        await applySettings(message.options, message.to_table, dynamoDB)
        await dynamoDB.deleteTable({ TableName: message.from_table })
        if (!message.to_table.includes('-restoring')) {
          message = { ...message, status: 'COMPLETED' }
        } else {
          message = { ...message, status: 'DELETING_ORIGINAL_TABLE' }
        }
      }
      await sqs.sendMessage({ QueueUrl: process.env['SQS_URL'], MessageBody: JSON.stringify(message) }).promise()
    }
  } catch (e) {
    throw Error(`Something went wrong when consuming your restore messages: ${e.message}`)
  }
}

const applySettings = async (options: StreamSpecification, toTable: string, dynamoDB: DynamoDB): Promise<void> => {
  try {
    await dynamoDB
      .updateTable({
        TableName: toTable,
        StreamSpecification: options,
      })
      .promise()

    await dynamoDB
      .updateContinuousBackups({
        TableName: toTable,
        PointInTimeRecoverySpecification: { PointInTimeRecoveryEnabled: true },
      })
  } catch (e) {
    throw Error(`Unable to apply settings to ${toTable}: ${e.message}`)
  }
}

const tableExists = async (tableName: string, dynamoDB: DynamoDB): Promise<boolean> => {
  try {
    const response = await dynamoDB.listTables().promise()
    return response.TableNames?.includes(tableName)!
  } catch (e) {
    throw Error(`Something happened when checking if ${tableName} is active: ${e.message}`)
  }
}